api/oc_push.c

```diff
@@ -0,0 +1,2605 @@
+static oc_resource_t *get_ns_instance(const char *href, oc_string_array_t *types,
+								oc_resource_properties_t bm, oc_interface_mask_t iface_mask,
+								size_t device)
+{
+	oc_ns_t *ns_instance = (oc_ns_t *)oc_memb_alloc(&ns_instance_memb);
+
+	if (ns_instance) {
+		ns_instance->resource = oc_new_resource(NULL, href, oc_string_array_get_allocated_size(*types), device);
+		if (ns_instance->resource) {
+			for (int i = 0; i < (int)oc_string_array_get_allocated_size(*types); i++) {
+				const char *rt = oc_string_array_get_item(*types, i);
+				oc_resource_bind_resource_type(ns_instance->resource, rt);
+			}
+			oc_resource_bind_resource_interface(ns_instance->resource, iface_mask);
+			ns_instance->resource->properties = bm;
+			oc_resource_set_default_interface(ns_instance->resource, OC_IF_RW);
+			oc_resource_set_request_handler(ns_instance->resource, OC_GET, get_ns, ns_instance);
+			oc_resource_set_request_handler(ns_instance->resource, OC_POST, post_ns, ns_instance);
+			oc_resource_set_request_handler(ns_instance->resource, OC_DELETE, delete_ns, ns_instance);
+			oc_resource_set_properties_cbs(ns_instance->resource, get_ns_properties, ns_instance, set_ns_properties, ns_instance);
+			oc_add_resource(ns_instance->resource);
+
+			OC_PUSH_DBG("new link (\"%s\") and corresponding resource for \"%s\" collection is created", oc_string(ns_instance->resource->uri), PUSHCONF_RSC_PATH);
+
+			oc_new_string(&ns_instance->state, pp_statestr(OC_PP_WFP), strlen(pp_statestr(OC_PP_WFP)));
+
+			OC_PUSH_DBG("state of Push Proxy (\"%s\") is initialized (%s)", oc_string(ns_instance->resource->uri), pp_statestr(OC_PP_WFP));
+
+			/*
+			 * add this new Notification Selector Resource to the list
+			 * which keeps all Notification Selectors of all Devices
+			 */
+			oc_list_add(ns_list, ns_instance);
+			return ns_instance->resource;
+		} else {
+			OC_PUSH_ERR("oc_new_resource() error!");
+			oc_memb_free(&ns_instance_memb, ns_instance);
+		}
+	}
+	else
+	{
+		OC_PUSH_ERR("oc_memb_alloc() error!");
+	}
+
+	return NULL;
+}
+
+
+/*
+ *
+ * @brief callback for freeing existing notification selector
+ * 		(this callback is called when target resource pointed by `link` is deleted by calling `oc_delete_resource()`)
+ *
+ */
+static void free_ns_instance(oc_resource_t *resource)
+{
+	oc_ns_t *ns_instance = (oc_ns_t *)oc_list_head(ns_list);
+	oc_endpoint_t *ep;
+
+	OC_PUSH_DBG("delete ns_instance for resource (\"%s\")...", oc_string(resource->uri));
+
+	while (ns_instance)
+	{
+		if (ns_instance->resource == resource)
+		{
+			/* remove link target resource itself here... */
+			oc_delete_resource(resource);
+
+			/* remove oc_ns_t instance from list */
+			oc_list_remove(ns_list, ns_instance);
+
+			/* free each field of ns_instance */
+			oc_free_string(&ns_instance->phref);
+			oc_free_string_array(&ns_instance->prt);
+			oc_free_string_array(&ns_instance->pif);
+
+			ep = ns_instance->pushtarget_ep.next;
+			while (ep)
+			{
+				oc_free_endpoint(ep);
```

Comment:
`oc_free_endpoint` deallocates all memory, including `next`, so you have to use a helper variable to avoid use of memory after free:

```C
  oc_endpoint_t *next;
  while (ep != NULL) {
    next = ep->next;
    oc_free_endpoint(ep);
    ep = next;
  }
```

Reply:
oops.. fixed!